File: coauthor/housekeeping/latexdiff.py
import os
from datetime import datetime

from ..utils.file import delete_file, move_file, find_file
from ..logger import logger

from .constants import TEMP_EXTENSIONS


def run_pack_latexdiff_vc(inputFile: str, commitHash: str, clean: bool = False) -> None:
    """Pack or clean latexdiff-vc output files into timestamped directory or remove them."""
    baseName = os.path.splitext(os.path.basename(inputFile))[0]
    input_dir = os.path.dirname(inputFile)
    output_folder = None if clean else os.path.join(input_dir, "Diffs", f"{datetime.now().strftime('%Y%m%d%H%M')}_{baseName}_{commitHash}")

    file_patterns = [f"{baseName}-diff{commitHash}"]
    files_to_process = []
    files_to_delete = []

    for pattern in file_patterns:
        for ext in [".tex", ".pdf"]:
            filePath = find_file(input_dir, pattern, ext)
            if filePath:
                files_to_process.append(filePath)
                for temp_ext in TEMP_EXTENSIONS:
                    temp_file = os.path.splitext(filePath)[0] + temp_ext
                    if os.path.exists(temp_file):
                        files_to_delete.append(temp_file)

    if files_to_process:
        if clean:
            for filePath in files_to_process + files_to_delete:
                delete_file(filePath)
            logger.info("Cleanup finished")
        else:  # move files to output folder
            os.makedirs(output_folder, exist_ok=True)
            for filePath in files_to_process:
                move_file(filePath, output_folder)

            for filePath in files_to_delete:
                delete_file(filePath)

            logger.info(f"Files packed to: {output_folder}")
    else:
        logger.warning("No files found to process.")


def run_pack_latexdiff_vc_multiple(inputFiles: list[str], commitHash: str, clean: bool = False) -> None:
    """Pack or clean latexdiff-vc output files for multiple input files."""
    for inputFile in inputFiles:
        run_pack_latexdiff_vc(inputFile, commitHash, clean)

File: src/housekeeping/latexdiff.ts
import * as path from 'node:path';

import * as logger from '@logger/logUtils';
import { WorkspaceFS } from '@utils/files';
import { toErrorMessage } from '@utils/errors/errorMessage';

import { findFilesFromPatterns, generateTimestamp } from './utils';
import { TEMP_EXTENSIONS } from './constants';

const CHANNEL = 'Housekeeping';

export type LatexdiffPackResult =
  | {
      status: 'no-files';
      inputFile: string;
    }
  | {
      status: 'cleaned';
      inputFile: string;
    }
  | {
      status: 'packed';
      inputFile: string;
      outputFolder: string;
    }
  | {
      status: 'processed';
      inputFile: string;
    }
  | {
      status: 'missing-inputs';
    }
  | {
      status: 'error';
      inputFile: string;
      error: unknown;
    };

export async function runPackLatexdiffvc(
  inputFile: string,
  commitHash: string,
  clean: boolean = false,
): Promise<LatexdiffPackResult> {
  const baseName = path.parse(inputFile).name;
  const inputDir = path.dirname(inputFile);
  const filePatterns = [`${baseName}-diff${commitHash}`];

  const mainFiles = findFilesFromPatterns(inputDir, filePatterns, [
    '.tex',
    '.pdf',
  ]);
  const tempFiles = findFilesFromPatterns(
    inputDir,
    filePatterns,
    TEMP_EXTENSIONS,
  );

  if (mainFiles.length === 0 && tempFiles.length === 0) {
    logger.warn(CHANNEL, 'No LaTeX diff files found to process');
    return { status: 'no-files', inputFile };
  }

  if (clean) {
    for (const file of [...mainFiles, ...tempFiles]) {
      await WorkspaceFS.delete(file);
    }
    logger.info(CHANNEL, 'Cleanup complete.');
    return { status: 'cleaned', inputFile };
  }

  const now = generateTimestamp();
  const outputFolder = path.join(
    inputDir,
    'Diffs',
    `${now}_${baseName}_${commitHash}`,
  );

  let dirCreated = false;
  for (const file of mainFiles) {
    if (!dirCreated) {
      await WorkspaceFS.createDir(outputFolder);
      dirCreated = true;
    }
    await WorkspaceFS.rename(
      file,
      path.join(outputFolder, path.basename(file)),
    );
  }

  for (const file of tempFiles) {
    await WorkspaceFS.delete(file);
  }

  if (dirCreated) {
    logger.info(CHANNEL, `Files packed into ${outputFolder}`);
    return { status: 'packed', inputFile, outputFolder };
  }

  return { status: 'processed', inputFile };
}

export async function runPackLatexdiffvcMultiple(
  inputFiles: string[],
  commitHash: string,
  clean: boolean = false,
): Promise<LatexdiffPackResult[]> {
  if (!inputFiles || inputFiles.length === 0) {
    logger.error(
      CHANNEL,
      'No input files provided for multiple LaTeX diff packing',
    );
    return [{ status: 'missing-inputs' }];
  }

  const results: LatexdiffPackResult[] = [];
  for (const inputFile of inputFiles) {
    try {
      results.push(await runPackLatexdiffvc(inputFile, commitHash, clean));
    } catch (err) {
      logger.error(
        CHANNEL,
        `Error during packing ${inputFile}: ${toErrorMessage(err)}`,
      );
      results.push({ status: 'error', inputFile, error: err });
    }
  }
  return results;
}
